src/lib.rs

```diff
@@ -1791,40 +1801,31 @@ impl Connection {
             {
                 let frame = frame::Frame::HandshakeDone;
 
-                payload_len += frame.wire_len();
-                left -= frame.wire_len();
-
-                frames.push(frame);
-
-                self.handshake_done_sent = true;
+                if add_frame! {frames, frame, payload_len, left} {
+                    self.handshake_done_sent = true;
+                }
             }
 
             // Create MAX_STREAMS_BIDI frame.
             if self.streams.should_update_max_streams_bidi() {
                 let max = self.streams.update_max_streams_bidi();
                 let frame = frame::Frame::MaxStreamsBidi { max };
 
-                payload_len += frame.wire_len();
-                left -= frame.wire_len();
-
-                frames.push(frame);
-
-                ack_eliciting = true;
-                in_flight = true;
+                if add_frame! {frames, frame, payload_len, left} {
+                    ack_eliciting = true;
+                    in_flight = true;
+                }
             }
 
             // Create MAX_STREAMS_UNI frame.
             if self.streams.should_update_max_streams_uni() {
                 let max = self.streams.update_max_streams_uni();
                 let frame = frame::Frame::MaxStreamsUni { max };
 
-                payload_len += frame.wire_len();
-                left -= frame.wire_len();
-
-                frames.push(frame);
-
-                ack_eliciting = true;
-                in_flight = true;
+                if add_frame! {frames, frame, payload_len, left} {
```

Comment:
Same as `MAX_STREAMS_BIDI`.

Reply:
See above for update_max_streams_uni() case.

Reply:
Also did the same for stream.max_data().